text/0000-panic-in-drop.md

```diff
@@ -0,0 +1,319 @@
+    some_function_that_may_panic();
+}
+
+fn expanded() {
+    let a = Box::new(1);
+    let b = try {
+        Box::new(2)
+    } catch {
+        try {
+            drop(a);
+        } catch {
+            abort();
+        }
+        resume_unwind();
+    };
+
+    try {
+        some_function_that_may_panic();
+    } catch {
+        try {
+            drop(a);
+            drop(b);
+        } catch {
+            abort();
+        }
+        resume_unwind();
+    }
+
+    try {
+        drop(a);
+    } catch {
+        try {
+            drop(b);
+        } catch {
+            abort();
+        }
+        resume_unwind();
+    }
+
+    drop(b);
+}
+
+fn expanded_without_drop_unwind() {
+    let a = Box::new(1);
+    let b = try {
+        Box::new(2)
+    } catch {
+        drop(a);
+        resume_unwind();
+    };
+
+    try {
+        some_function_that_may_panic();
+    } catch {
+        drop(a);
+        drop(b);
+        resume_unwind();
+    }
+
+    drop(a);
+    drop(b);
+}
+```
+
+Even if the exception safety concerns can be dismissed as "just buggy unsafe code", the code size and compilation time benefits alone should justify making this change.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC changes how panics behave when called from within a `Drop` implementation. Previously, a panicking `Drop` would behave differently depending on whether another panic was currently in the process of unwinding the stack.
+
+- If it was, then the process was immediately aborted with a double-panic error.
+- Otherwise, the panic would:
+  - Skip the rest of the `Drop` impl.
+  - Continue dropping the fields of the object with the panicking `Drop` impl.
+  - If that object was itself part of a `struct`, array or `Vec` then the remaining fields/array elements are also dropped.
+
+This behavior was non-obvious and error-prone. It has been the source of several security bugs in unsafe code both [inside](https://github.com/rust-lang/rust/issues/83618) and [outside](https://github.com/bluss/arrayvec/issues/3) the standard library.
+
+With this RFC the behavior of panics in `Drop` is changed to always abort the process. This mirrors a [similar change](https://akrzemi1.wordpress.com/2013/08/20/noexcept-destructors/) made in C++11 where destructors terminate the process if an exception escapes them.
+
+If you have any existing code which relies on panics in `Drop` not aborting the process, you should refactor your code to return errors using a method that consumes `self`. For example consider this hypothetical `close` method on `BufWriter`:
+
+```rust
+impl BufWriter {
+    /// Closes the underlying file and reports any errors that occurred while
+    /// flushing the buffer.
+    ///
+    /// Note that the underlying file is still closed even if I/O errors
+    /// occurred.
+    pub fn close(self) -> std::io::Result<()> {
+        let err = self.flush();
+        drop(self);
+        err
+    }
+}
+```
```

Comment:
This RFC might not be the exact place for it, but I would rather the example here show the theoretical `Drop` implementation as well. *Not* flushing on `drop` would be a *horrible* regression in RAII ergonomics, but a naive implementation like this results in a double flush. The double flush isn't *that* problematic in this case, but if a type was using a deliberately panicking destructor previously it was likely that a double flush is more problematic in that case.

<details><summary>naive implementation</summary>

```rust
impl BufWriter {
    pub fn close(self) -> std::io::Result<()> {
        let err = self.flush();
        drop(self);
        err
    }
}

impl Drop for BufWriter {
    fn drop(&mut self) {
        let err = self.flush();
        drop(err);
    }
}
```

</details>

<details>